README.md
# DEPRECATED

This is no longer supported. Its original contents has been moved to other projects.

* `Principal`: [`candid`](https://github.com/dfinity/candid/blob/28a9fe8e898dc7662de2a13204b6f06bce536e14/rust/candid/src/types/principal.rs).
* `HashTree`: [`agent-rs`](https://github.com/dfinity/agent-rs/tree/c9f9497dc8ced64b76250d76967f4f61a56d8262/ic-certification).

# IC Types

## Contributing
Please follow the guidelines in the [CONTRIBUTING.md](.github/CONTRIBUTING.md) document.

## Goal
This library contains typings and utility functions dealing with the Public specification of the Internet Computer
 and the HTTP
client. It might be shared in the future but for now is separated for the purpose of testing and
development.

## Running Tests
Regular tests can be run by anyone using `cargo test`. This will not run a special version of the
tests that runs against a server implementation of the public specification, though those
tests can be run against a compatible Replica.

## References
See https://sdk.dfinity.org/docs/interface-spec/index.html


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.7.0] - 2022-11-22

- Deprecate this crate.

## [0.6.0] - 2022-09-28

- Move out `Principal` to `candid` crate.

## [0.5.0] - 2022-09-22

### Changed
- Make the function `lookup_path` and `lookup_label` comply with Interface Spec.

## [0.4.2] - 2022-08-30

### Changed
- Add a function `list_paths` to list all paths containing values in a hash tree.

## [0.4.1] - 2022-07-15

### Changed
- Remove deprecation of `Principal::from_slice`.

## [0.4.0] - 2022-07-13

### Changed
- `PrincipalError` enum has different set of variants reflecting changes in `Principal::from_text` logic.
- `Principal::from_text` accepts input containing uppercase letters which results in Err before.
- `Principal::from_text` verifies CRC32 check sequence.
- `Principal::from_slice` is deprecated in favor of `Principal::try_from_slice`.

### Fixed
- Converting long bytes ending in `0x04` to `Principal` is ok.

## [0.3.0] - 2022-01-04

### Changed
- lookup_path uses an Iterator rather than an array.
- Principal, Label, and HashTree implement From instead of Into in order to remove needless borrowing.


tests/principal.rs
#![allow(deprecated)]

use ic_types::Principal;
use ic_types::PrincipalError;

const MANAGEMENT_CANISTER_BYTES: [u8; 0] = [];
const MANAGEMENT_CANISTER_TEXT: &str = "aaaaa-aa";

const ANONYMOUS_CANISTER_BYTES: [u8; 1] = [4u8];
const ANONYMOUS_CANISTER_TEXT: &str = "2vxsx-fae";

const TEST_CASE_BYTES: [u8; 9] = [0xef, 0xcd, 0xab, 0, 0, 0, 0, 0, 1];
const TEST_CASE_TEXT: &str = "2chl6-4hpzw-vqaaa-aaaaa-c";

mod convert_from_bytes {
    use super::*;

    #[test]
    fn try_from_test_case_ok() {
        assert!(Principal::try_from_slice(&TEST_CASE_BYTES).is_ok());
    }

    #[test]
    fn try_from_slice_length_0_29_ok() {
        let array = [0u8; 29];
        for len in 0..30 {
            assert!(Principal::try_from_slice(&array[..len]).is_ok());
        }
    }

    #[test]
    fn try_from_slice_length_30_err() {
        assert_eq!(
            Principal::try_from_slice(&[0u8; 30]),
            Err(PrincipalError::BytesTooLong())
        );
    }

    #[test]
    fn from_test_case_ok() {
        Principal::from_slice(&TEST_CASE_BYTES);
    }

    #[test]
    fn from_slice_length_0_29_ok() {
        let array = [0u8; 29];
        for len in 0..30 {
            Principal::from_slice(&array[..len]);
        }
    }

    #[test]
    #[should_panic]
    fn from_slice_length_30_err() {
        Principal::from_slice(&[0u8; 30]);
    }
}

mod convert_from_text {
    use super::*;

    #[test]
    fn convert_from_test_case_ok() {
        // input must be 8~63 chars including `-`s
        // using empty blob as shortest test case
        assert!(Principal::from_text(TEST_CASE_TEXT).is_ok());
    }

    #[test]
    fn convert_from_management_canister_text_ok() {
        // input must be 8~63 chars including `-`s
        // using empty blob as shortest test case
        assert!(Principal::from_text(MANAGEMENT_CANISTER_TEXT).is_ok());
    }

    #[test]
    fn convert_from_29_0u8_ok() {
        // input must be 8~63 chars including `-`s
        // using [0u8; 29] as longest test case
        assert!(Principal::from_text(
            "bnkmk-jqaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaa"
        )
        .is_ok());
    }

    #[test]
    fn convert_from_6a_invalid_base_32() {
        assert_eq!(
            Principal::from_text("aaaaa-a"),
            Err(PrincipalError::InvalidBase32())
        );
    }

    #[test]
    fn convert_from_5a_text_too_short() {
        assert_eq!(
            Principal::from_text("aaaaa"),
            Err(PrincipalError::TextTooShort())
        );
    }

    #[test]
    fn convert_from_30_0u8_text_too_long() {
        assert_eq!(
            Principal::from_text(
                "aacd5-niaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaaa-aaaaa"
            ),
            Err(PrincipalError::TextTooLong())
        );
    }

    #[test]
    fn convert_from_wrong_dash_position_err() {
        assert_eq!(
            Principal::from_text("aaaaaaa"),
            Err(PrincipalError::AbnormalGrouped(
                Principal::management_canister()
            ))
        );
        assert_eq!(
            Principal::from_text("aaaaaa-a"),
            Err(PrincipalError::AbnormalGrouped(
                Principal::management_canister()
            ))
        );
        assert_eq!(
            Principal::from_text("aaaaaaa-"),
            Err(PrincipalError::AbnormalGrouped(
                Principal::management_canister()
            ))
        );
        assert_eq!(
            Principal::from_text("-aaaaaaa"),
            Err(PrincipalError::AbnormalGrouped(
                Principal::management_canister()
            ))
        );
        assert_eq!(
            Principal::from_text("aaa-aaaa"),
            Err(PrincipalError::AbnormalGrouped(
                Principal::management_canister()
            ))
        );
    }

    #[test]
    fn convert_from_capitals_ok() {
        assert_eq!(
            Principal::from_text("AAAAA-AA"),
            Ok(Principal::management_canister())
        );
        assert_eq!(
            Principal::from_text("aaaaa-AA"),
            Ok(Principal::management_canister())
        );
    }
}

mod convert_to_bytes {
    use super::*;

    #[test]
    fn managements_canister_converts_to_empty_slice() {
        assert_eq!(
            Principal::management_canister().as_slice(),
            &MANAGEMENT_CANISTER_BYTES
        );
    }

    #[test]
    fn anonymouse_converts_to_single_byte_04() {
        assert_eq!(Principal::anonymous().as_slice(), &ANONYMOUS_CANISTER_BYTES);
    }

    #[test]
    fn test_case_to_bytes_correct() {
        assert_eq!(
            Principal::from_text(TEST_CASE_TEXT).unwrap().as_slice(),
            &TEST_CASE_BYTES
        );
    }
}

mod convert_to_text {
    use super::*;

    #[test]
    fn managements_canister_to_text_correct() {
        assert_eq!(
            Principal::management_canister().to_text(),
            MANAGEMENT_CANISTER_TEXT.to_string()
        );
    }

    #[test]
    fn anonymous_to_text_correct() {
        assert_eq!(
            Principal::anonymous().to_text(),
            ANONYMOUS_CANISTER_TEXT.to_string()
        );
    }

    #[test]
    fn test_case_to_text_correct() {
        assert_eq!(
            Principal::try_from_slice(&TEST_CASE_BYTES)
                .unwrap()
                .to_text(),
            TEST_CASE_TEXT.to_string()
        );
    }
}

mod ser_de {
    use super::*;
    use serde_test::{assert_tokens, Configure, Token};

    #[test]
    fn management_canister_serde_match() {
        let p = Principal::management_canister();
        assert_tokens(&p.compact(), &[Token::Bytes(&MANAGEMENT_CANISTER_BYTES)]);
        assert_tokens(&p.readable(), &[Token::Str(MANAGEMENT_CANISTER_TEXT)]);
    }

    #[test]
    fn test_case_serde_match() {
        let p = Principal::try_from_slice(&TEST_CASE_BYTES).unwrap();
        assert_tokens(&p.compact(), &[Token::Bytes(&TEST_CASE_BYTES)]);
        assert_tokens(&p.readable(), &[Token::Str(TEST_CASE_TEXT)]);
    }
}

#[test]
fn impl_traits() {
    use serde::{Deserialize, Serialize};
    use std::convert::TryFrom;
    use std::fmt::{Debug, Display};
    use std::hash::Hash;
    use std::str::FromStr;

    assert!(impls::impls!(
        Principal: Debug & Display & Clone & Copy & Eq & PartialOrd & Ord & Hash
    ));

    assert!(
        impls::impls!(Principal: FromStr & TryFrom<&'static str> & TryFrom<Vec<u8>> & TryFrom<&'static Vec<u8>> & TryFrom<&'static [u8]> & AsRef<[u8]>)
    );

    assert!(impls::impls!(Principal: Serialize & Deserialize<'static>));
}

#[test]
fn long_blobs_ending_04_is_valid_principal() {
    let blob: [u8; 18] = [
        10, 116, 105, 100, 0, 0, 0, 0, 0, 144, 0, 51, 1, 1, 0, 0, 0, 4,
    ];
    assert!(Principal::try_from_slice(&blob).is_ok());
}

#[test]
fn self_authenticating_ok() {
    // self_authenticating doesn't verify the input bytes
    // this test checks:
    // 1. Sha224 hash is used
    // 2. 0x02 was added in the end
    // 3. total length is 29
    let p1 = Principal::self_authenticating(&[]);
    let p2 = Principal::try_from_slice(&[
        209, 74, 2, 140, 42, 58, 43, 201, 71, 97, 2, 187, 40, 130, 52, 196, 21, 162, 176, 31, 130,
        142, 166, 42, 197, 179, 228, 47, 2,
    ])
    .unwrap();
    assert_eq!(p1, p2);
}


src/hash_tree/hash_tree_tests.rs
#![cfg(test)]
use crate::hash_tree::{
    empty, fork, label, leaf, pruned, pruned_from_hex, HashTree, Label, LookupResult,
};

fn lookup_path<'a, P: AsRef<[&'static str]>>(tree: &'a HashTree<'a>, path: P) -> LookupResult<'a> {
    let path: Vec<Label> = path.as_ref().iter().map(|l| l.into()).collect();

    tree.lookup_path(&path)
}

#[test]
fn works_with_simple_tree() {
    let tree = fork(
        label("label 1", empty()),
        fork(
            pruned(*b"\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01"),
            leaf(&[1u8, 2, 3, 4, 5, 6]),
        ),
    );

    assert_eq!(
        hex::encode(tree.digest().to_vec()),
        "69cf325d0f20505b261821a7e77ff72fb9a8753a7964f0b587553bfb44e72532"
    );
}

#[test]
fn spec_example() {
    // This is the example straight from the spec.
    let tree = fork(
        fork(
            label(
                "a",
                fork(
                    fork(label("x", leaf(b"hello")), empty()),
                    label("y", leaf(b"world")),
                ),
            ),
            label("b", leaf(b"good")),
        ),
        fork(label("c", empty()), label("d", leaf(b"morning"))),
    );

    // Check CBOR serialization.
    #[cfg(feature = "serde")]
    assert_eq!(
        hex::encode(serde_cbor::to_vec(&tree).unwrap()),
        "8301830183024161830183018302417882034568656c6c6f810083024179820345776f726c6483024162820344676f6f648301830241638100830241648203476d6f726e696e67"
    );

    assert_eq!(
        hex::encode(tree.digest().to_vec()),
        "eb5c5b2195e62d996b84c9bcc8259d19a83786a2f59e0878cec84c811f669aa0"
    );
}

#[test]
fn spec_example_pruned() {
    // This is the example straight from the spec.
    let tree = fork(
        fork(
            label(
                "a",
                fork(
                    pruned_from_hex(
                        "1b4feff9bef8131788b0c9dc6dbad6e81e524249c879e9f10f71ce3749f5a638",
                    )
                    .unwrap(),
                    label("y", leaf(b"world")),
                ),
            ),
            label(
                "b",
                pruned_from_hex("7b32ac0c6ba8ce35ac82c255fc7906f7fc130dab2a090f80fe12f9c2cae83ba6")
                    .unwrap(),
            ),
        ),
        fork(
            pruned_from_hex("ec8324b8a1f1ac16bd2e806edba78006479c9877fed4eb464a25485465af601d")
                .unwrap(),
            label("d", leaf(b"morning")),
        ),
    );

    assert_eq!(
        hex::encode(tree.digest().to_vec()),
        "eb5c5b2195e62d996b84c9bcc8259d19a83786a2f59e0878cec84c811f669aa0"
    );

    assert_eq!(lookup_path(&tree, ["a", "a"]), LookupResult::Unknown);
    assert_eq!(
        lookup_path(&tree, ["a", "y"]),
        LookupResult::Found(b"world")
    );
    assert_eq!(lookup_path(&tree, ["aa"]), LookupResult::Absent);
    assert_eq!(lookup_path(&tree, ["ax"]), LookupResult::Absent);
    assert_eq!(lookup_path(&tree, ["b"]), LookupResult::Unknown);
    assert_eq!(lookup_path(&tree, ["bb"]), LookupResult::Unknown);
    assert_eq!(lookup_path(&tree, ["d"]), LookupResult::Found(b"morning"));
    assert_eq!(lookup_path(&tree, ["e"]), LookupResult::Absent);
}

#[test]
fn can_lookup_paths_1() {
    let tree = fork(
        label("label 1", empty()),
        fork(
            pruned([1; 32]),
            fork(
                label("label 3", leaf(vec![1, 2, 3, 4, 5, 6])),
                label("label 5", empty()),
            ),
        ),
    );

    assert_eq!(
        tree.lookup_path(&["label 0".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 1".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 2".into()]),
        LookupResult::Unknown
    );
    assert_eq!(
        tree.lookup_path(&["label 3".into()]),
        LookupResult::Found(&[1, 2, 3, 4, 5, 6])
    );
    assert_eq!(
        tree.lookup_path(&["label 4".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 5".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 6".into()]),
        LookupResult::Absent
    );
}

#[test]
fn can_lookup_paths_2() {
    let tree = fork(
        label("label 1", empty()),
        fork(
            fork(
                label("label 3", leaf(vec![1, 2, 3, 4, 5, 6])),
                label("label 5", empty()),
            ),
            pruned([1; 32]),
        ),
    );

    assert_eq!(
        tree.lookup_path(&["label 0".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 1".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 2".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 3".into()]),
        LookupResult::Found(&[1, 2, 3, 4, 5, 6])
    );
    assert_eq!(
        tree.lookup_path(&["label 4".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 5".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 6".into()]),
        LookupResult::Unknown
    );
}

#[test]
fn can_lookup_paths_3() {
    let tree = fork(
        pruned([0; 32]),
        fork(
            pruned([1; 32]),
            fork(
                label("label 3", leaf(vec![1, 2, 3, 4, 5, 6])),
                label("label 5", empty()),
            ),
        ),
    );

    assert_eq!(
        tree.lookup_path(&["label 2".into()]),
        LookupResult::Unknown
    );
    assert_eq!(
        tree.lookup_path(&["label 3".into()]),
        LookupResult::Found(&[1, 2, 3, 4, 5, 6])
    );
    assert_eq!(
        tree.lookup_path(&["label 4".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 5".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 6".into()]),
        LookupResult::Absent
    );
}

#[test]
fn can_lookup_paths_4() {
    let tree = fork(
        pruned([0; 32]),
        fork(
            fork(
                label("label 3", leaf(vec![1, 2, 3, 4, 5, 6])),
                label("label 5", empty()),
            ),
            pruned([1; 32]),
        ),
    );

    assert_eq!(
        tree.lookup_path(&["label 2".into()]),
        LookupResult::Unknown
    );
    assert_eq!(
        tree.lookup_path(&["label 3".into()]),
        LookupResult::Found(&[1, 2, 3, 4, 5, 6])
    );
    assert_eq!(
        tree.lookup_path(&["label 4".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 5".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 6".into()]),
        LookupResult::Unknown
    );
}

#[test]
fn can_lookup_paths_5() {
    let tree = fork(
        fork(
            pruned([1; 32]),
            fork(
                label("label 3", leaf(vec![1, 2, 3, 4, 5, 6])),
                label("label 5", empty()),
            ),
        ),
        label("label 7", empty()),
    );

    assert_eq!(
        tree.lookup_path(&["label 2".into()]),
        LookupResult::Unknown
    );
    assert_eq!(
        tree.lookup_path(&["label 3".into()]),
        LookupResult::Found(&[1, 2, 3, 4, 5, 6])
    );
    assert_eq!(
        tree.lookup_path(&["label 4".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 5".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 6".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 7".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 8".into()]),
        LookupResult::Absent
    );
}

#[test]
fn can_lookup_paths_6() {
    let tree = fork(
        fork(
            fork(
                label("label 3", leaf(vec![1, 2, 3, 4, 5, 6])),
                label("label 5", empty()),
            ),
            pruned([1; 32]),
        ),
        label("label 7", empty()),
    );

    assert_eq!(
        tree.lookup_path(&["label 2".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 3".into()]),
        LookupResult::Found(&[1, 2, 3, 4, 5, 6])
    );
    assert_eq!(
        tree.lookup_path(&["label 4".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 5".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 6".into()]),
        LookupResult::Unknown
    );
    assert_eq!(
        tree.lookup_path(&["label 7".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 8".into()]),
        LookupResult::Absent
    );
}

#[test]
fn can_lookup_paths_7() {
    let tree = fork(
        fork(
            pruned([1; 32]),
            fork(
                label("label 3", leaf(vec![1, 2, 3, 4, 5, 6])),
                label("label 5", empty()),
            ),
        ),
        pruned([0; 32]),
    );

    assert_eq!(
        tree.lookup_path(&["label 2".into()]),
        LookupResult::Unknown
    );
    assert_eq!(
        tree.lookup_path(&["label 3".into()]),
        LookupResult::Found(&[1, 2, 3, 4, 5, 6])
    );
    assert_eq!(
        tree.lookup_path(&["label 4".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 5".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 6".into()]),
        LookupResult::Unknown
    );
}

#[test]
fn can_lookup_paths_8() {
    let tree = fork(
        fork(
            fork(
                label("label 3", leaf(vec![1, 2, 3, 4, 5, 6])),
                label("label 5", empty()),
            ),
            pruned([1; 32]),
        ),
        pruned([0; 32]),
    );

    assert_eq!(
        tree.lookup_path(&["label 2".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 3".into()]),
        LookupResult::Found(&[1, 2, 3, 4, 5, 6])
    );
    assert_eq!(
        tree.lookup_path(&["label 4".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 5".into()]),
        LookupResult::Absent
    );
    assert_eq!(
        tree.lookup_path(&["label 6".into()]),
        LookupResult::Unknown
    );
}


src/blob.rs
use std::fmt;
use crate::utils::byte_slice_fmt::truncate_and_format;

/// A binary "blob", i.e. a byte array.
///
/// Use `serde_bytes` so that the `Vec<u8>` is deserialized as a sequence
/// (array) of bytes, whereas we want an actual CBOR "byte array", e.g. a
/// byte string.
#[derive(Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Blob(#[cfg_attr(feature = "serde", serde(with = "serde_bytes"))] pub Vec<u8>);

impl Blob {
    fn format(&self, f: &mut fmt::Formatter<'_>, max_bytes_to_format: usize) -> fmt::Result {
        f.write_fmt(format_args!(
            "Blob{{{}}}",
            truncate_and_format(self.0.as_slice(), max_bytes_to_format)
        ))
    }
}

impl fmt::Debug for Blob {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.format(f, usize::MAX)
    }
}

impl fmt::Display for Blob {
    // Just like Debug, except we truncate long ones
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.format(f, 40_usize)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn serialize_to_cbor() {
        let bytes = vec![1, 2, 3, 4, 5, 6];
        let blob = Blob(bytes);

        let cbor_bytes = serde_cbor::to_vec(&blob).expect("Failed to encode blob to cbor");
        let cbor_hex = hex::encode(cbor_bytes);

        assert_eq!(cbor_hex, String::from("46010203040506"));
    }

    #[test]
    fn format() {
        let bytes = vec![1, 2, 3, 4, 5, 6];
        let blob = Blob(bytes);

        let formatted_blob = format!("{:?}", blob);

        assert_eq!(formatted_blob, "Blob{6 bytes;010203040506}");
    }

    #[test]
    fn format_large_blob() {
        let bytes = vec![1; 42];
        let blob = Blob(bytes);

        let formatted_blob = format!("{}", blob);

        assert_eq!(formatted_blob, "Blob{42 bytes;01010101010101010101010101010101010101010101010101010101010101010101010101010101…}");
    }

    #[test]
    fn format_large_blob_debug() {
        let bytes = vec![1; 42];
        let blob = Blob(bytes);

        let formatted_blob = format!("{:?}", blob);

        assert_eq!(formatted_blob, "Blob{42 bytes;010101010101010101010101010101010101010101010101010101010101010101010101010101010101}");
    }
}


src/utils/byte_slice_fmt.rs
/// Formats the first `max_bytes_to_format` bytes of a byte slice as
/// hexadecimal.
pub fn truncate_and_format(slice: &[u8], max_bytes_to_format: usize) -> String {
    let truncated = &slice[..slice.len().min(max_bytes_to_format)];
    let content_hex = hex::encode(&truncated);
    let ellipsis = if truncated.len() < slice.len() {
        "…"
    } else {
        ""
    };
    let size_descriptor = match slice.len() {
        0 => "empty".to_string(),
        1..=3 => String::default(), // For small amount of bytes, it's easy to count
        n => format!("{} bytes;", n),
    };
    format!("{}{}{}", size_descriptor, content_hex, ellipsis)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn f(vec: Vec<u8>, max_bytes_to_format: usize) -> String {
        truncate_and_format(vec.as_slice(), max_bytes_to_format)
    }

    #[test]
    fn test_truncate_and_format() {
        assert_eq!(f(vec![], 5), "empty");
        assert_eq!(f(vec![0], 5), "00");
        assert_eq!(f(vec![255, 0], 5), "ff00");
        assert_eq!(f(vec![1, 2, 3], 5), "010203");
        assert_eq!(f(vec![0, 1, 15, 255], 5), "4 bytes;00010fff");
        assert_eq!(f((0_u8..100_u8).collect(), 5), "100 bytes;0001020304…");
        assert_eq!(f((0_u8..100_u8).collect(), 6), "100 bytes;000102030405…");
    }
}


src/certificate.rs
use crate::{Blob, HashTree};

/// A `Certificate` as defined in https://internetcomputer.org/docs/current/references/ic-interface-spec/#certificate
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Certificate<'a> {
    pub tree: HashTree<'a>,
    pub signature: Blob,
    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
    pub delegation: Option<CertificateDelegation>,
}

/// A `CertificateDelegation` as defined in https://internetcomputer.org/docs/current/references/ic-interface-spec/#certification-delegation
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct CertificateDelegation {
    pub subnet_id: Blob,
    pub certificate: Blob,
}

#[cfg(test)]
mod tests {
    use crate::hash_tree::{empty, fork, label, leaf};
    use super::*;

    fn create_tree<'a>() -> HashTree<'a> {
        fork(
            fork(
                label(
                    "a",
                    fork(
                        fork(label("x", leaf(b"hello")), empty()),
                        label("y", leaf(b"world")),
                    ),
                ),
                label("b", leaf(b"good")),
            ),
            fork(label("c", empty()), label("d", leaf(b"morning"))),
        )
    }

    #[test]
    fn serialize_to_cbor() {
        let tree = create_tree();
        let signature = Blob(vec![1, 2, 3, 4, 5, 6]);

        let certificate = Certificate {
            tree,
            signature,
            delegation: None,
        };

        let cbor_bytes = serde_cbor::to_vec(&certificate).expect("Failed to encode certificate to cbor");
        let cbor_hex = hex::encode(cbor_bytes);

        assert_eq!(cbor_hex, "a264747265658301830183024161830183018302417882034568656c6c6f810083024179820345776f726c6483024162820344676f6f648301830241638100830241648203476d6f726e696e67697369676e617475726546010203040506");
    }

    #[test]
    fn serialize_to_cbor_with_delegation() {
        let tree = create_tree();
        let signature = Blob(vec![1, 2, 3, 4, 5, 6]);
        let delegation = CertificateDelegation {
            subnet_id: Blob(vec![7, 8, 9, 10, 11, 12]),
            certificate: Blob(vec![13, 14, 15, 16, 17, 18]),
        };

        let certificate = Certificate {
            tree,
            signature,
            delegation: Some(delegation),
        };

        let cbor_bytes = serde_cbor::to_vec(&certificate).expect("Failed to encode certificate to cbor");
        let cbor_hex = hex::encode(cbor_bytes);

        assert_eq!(cbor_hex, "a364747265658301830183024161830183018302417882034568656c6c6f810083024179820345776f726c6483024162820344676f6f648301830241638100830241648203476d6f726e696e67697369676e6174757265460102030405066a64656c65676174696f6ea2697375626e65745f6964460708090a0b0c6b6365727469666963617465460d0e0f101112");
    }
}
